### ISSUE TITLE
Implement ShowField Text tab

### ISSUE THREAD

**Issue by dcwhite** [SCIInstitute/SCIRun#2148]
For node/edge/face/data labeling, etc. Copy v4's UI.

**Comment by jessdtate** [comment]
look at showstring and showcolormap for the existing implementation.  SR4 implementation had the several field paramaters available for showing as text (field data, node index, etc) at the data location.  The text was projected and rotated to appear with next to the location even if moved.

**Comment by dcwhite** [comment]
**Analysis (2026-06-13)**

ShowField Text tab: the UI file has widgets but they're hidden in C++. v4 showed field data (values, node indices) as text labels projected at data locations in 3D. The renderer integration point would be ShowString + the spire text rendering path. Comment says to look at ShowString and ShowColorMap for the existing implementation. Moderate effort — the C++ side needs to unhide widgets and connect them to geometry generation code.

**Comment by dcwhite** [comment]
**Code analysis (2026-07-31)** — supersedes my 2026-06-13 note above, which understated this.

That earlier comment said "the C++ side needs to unhide widgets and connect them to geometry generation code." That's wrong in an important way: **the GUI is essentially complete, and there is nothing to connect it to.** This is a renderer task, not a UI task.

## What already exists

- `src/Interface/Modules/Visualization/ShowField.ui:682` — a complete `textTab_`: Show Text, per-element checkboxes, "Cull backfacing text", "Text always visible", Text Size, Text Precision, a Text Coloring radio group, and a color picker.
- `src/Interface/Modules/Visualization/ShowFieldDialog.cc:58-85` — every one of those widgets is already wired to module state (`addCheckBoxManager`, `addSpinBoxManager`, `addRadioButtonGroupManager`, plus `assignDefaultTextColor`).
- `src/Modules/Visualization/ShowField.cc:169-181` — all seven parameters (`ShowText`, `CullBackfacingText`, `TextAlwaysVisible`, `TextSize`, `TextPrecision`, `TextColoring`, `DefaultTextColor`) are declared and given defaults.

And **no code ever reads any of them.** The tab is removed outright at `ShowFieldDialog.cc:111`:

```cpp
//Text Tab
displayOptionsTabs_->removeTab(3);
```

This traces to `88f2161b8 "Revamped ShowField UI"` — the UI was built first and the backend never landed. Un-hiding the tab today just exposes seven controls that do nothing.

## The actual blockers — both structural, both in TextBuilder

### 1. The text path is screen-space only, by construction

`TextBuilder::printString` takes a position in normalized **screen** space (`src/Modules/Visualization/TextBuilder.h:60-66`), and the vertex shader at `src/Interface/Modules/Render/ES/shaders/TextBuilder.vs:41-49` computes `gl_Position` directly from pixel coordinates:

```glsl
gl_Position = vec4(aPos.x * x_scale + uTrans.x - 1.0,
                   aPos.y * y_scale + uTrans.y - 1.0,
                   -1.0, 1.0);
```

It never touches the model/view/projection matrices, and hardcodes `z = -1.0`. The v4 behavior described above — text at the data location, projected and rotating with the geometry — is simply not expressible in this path. It needs a new billboard shader: transform a world-space anchor through the MVP, then offset the glyph quad in screen space.

Related: text passes are routed away from `RenderBasicSys` (early return at `RenderBasicSys.cc:146` on `IS_TEXT`) and drawn by `RenderTransText` after everything else. So "Text always visible" is currently the *only* possible behavior, and "Cull backfacing text" has no surface normal to test against until the billboard path exists.

### 2. One draw call per character

The loop body at `src/Modules/Visualization/TextBuilder.cc:157-261` allocates a **VBO, an IBO, a texture, and a SubPass for every single glyph** — not per string, per character.

That is survivable for the only three callers today, all of which are tiny:
- `ShowString.cc:215` — one short string
- `ShowColorMapModule.cc:249` — a handful of tick labels
- `ViewScene.cc:2568` — axis labels

It is not survivable for ShowField. Labeling a modest 10k-node mesh with ~6 characters of field data per node is on the order of **60,000 draw calls and 60,000 textures**. The design has never been stressed because nothing has ever asked it to scale.

This needs a font atlas (one texture, glyph UVs) and a single batched VBO per label set.

## Estimate

Not conceptually hard — batched, atlased, billboarded text is well-trodden graphics work — but it is squarely renderer work, which is the opposite of how this issue currently reads.

| Piece | Rough effort |
|---|---|
| Font atlas + batched VBO in `TextBuilder` | 2–3 days |
| Billboard vertex shader (world anchor → MVP → screen-space glyph offset) | 1–2 days |
| Depth handling: real depth from the anchor, "always visible" toggle, backface cull via normal | 1 day |
| ShowField-side geometry generation: iterate nodes/edges/faces, format values (respecting Text Precision), apply the three Text Coloring modes | 2 days |
| Un-hide tab, `.srn5` state round-trip, regression net + golden image | 1 day |

**Total: roughly 1.5–2 weeks** of focused work for the real thing. Treat these as rough — the atlas rework is the item most likely to grow, since it touches all three existing callers.

A **~1 day proof-of-concept** is possible by reusing the existing per-character path and capping it to small meshes or a decimated node subset. That would validate the ShowField-side plumbing and the label formatting, but it would fall over on any realistic dataset and should not be mistaken for the feature.

## Worth noting

The atlas + shader work is shared infrastructure, not ShowField-specific. It would also:
- resolve #1334 (inconsistent and too-small font sizes across ShowString / ShowColorMap / ShowField), since a proper atlas makes arbitrary font sizing cheap, and
- meaningfully speed up ShowString and ShowColorMap by collapsing their per-character draw calls.

That argues for landing the TextBuilder rework as its own PR first, with the ShowField Text tab as the follow-up that consumes it.

**Comment by dcwhite** [comment]
## Does the VTK renderer make this moot?

**Partly — and specifically it moots the expensive half of my estimate above, but not the whole issue.** Which half depends on an architectural question that is currently unresolved (see "the crux" below). Worth deciding before anyone writes the font atlas.

### Where VTK actually stands in this repo (verified today, not assumed)

- **VTK is not a dependency at all.** Zero hits in `Superbuild/`. Compare OSPRay, which is fully wired (`Superbuild.cmake:106,243,283,315`, `src/CMakeLists.txt:305,391`). The only VTK code in the tree is file I/O (`VTKToTriSurfReader`), unrelated to rendering. #77 removed the old VTK rendering code back in 2012.
- Milestone [#98 "VTK renderer"](https://github.com/SCIInstitute/SCIRun/milestone/98): **0 closed, 3 open** (#2469, #2470, #2476). Due 2026-08-31, having already slipped from 2026-03-31.
- #2481 (abstract renderer interface) sits in a different milestone, *Refactor Visualization Layer*.
- No code on any of the four. All are analysis-only so far.

So in schedule terms: no, VTK does not make this moot in any near-term actionable sense. There is no VTK build, let alone a VTK renderer, and the two hardest risks flagged in #2469 (Qt/VTK GL context sharing, GeometryObject↔VTK buffer ownership) are both unstarted.

### One correction to the framing in #2481

`RendererInterface` **already exists** — `src/Interface/Modules/Render/ES/RendererInterface.h`, 145 lines, dating to 2020. So #2481 isn't a greenfield "design an abstraction" task; it's a *rework of a leaky existing one*. The current interface is written in the native renderer's own vocabulary — `glm::mat4`/`glm::quat`, `QOpenGLContext*`, `RenderState::TransparencySortType`, `StaticClippingPlanes*`, `es-general` transforms, even `hasShaderPromise()`. It's an extraction of `SRInterface`'s method list (its only implementer), not a renderer-neutral contract.

The one genuinely neutral seam in it is:

```cpp
virtual void handleGeomObject(Graphics::Datatypes::GeometryHandle object, int port) = 0;
```

**And relevant here: `RendererInterface` has no concept of text whatsoever.** Text bypasses it entirely — `TextBuilder` is called directly inside the ES subsystem. So text is not something a VTK backend would inherit for free from the existing abstraction; it has to be built on whichever side ends up owning it.

### The crux: fork or replace?

Milestone #98's own description leaves this explicitly open — *"run in parallel to or instead of the native SR version."* The two readings have **opposite** consequences for this issue:

| If VTK arrives as… | Consequence for #2148 |
|---|---|
| **A parallel module** (`ShowFieldVtk` #2470 + `ViewSceneVtk` #2469), following the `ShowFieldWithOspray` precedent | **Not moot at all.** Native `ShowField` never touches VTK, so it still needs its own text path — atlas, billboard, the works. Users on the default renderer get nothing. |
| **A true backend** behind a reworked `RendererInterface` (#2481), with `ShowField` delegating | **Largely moot.** The renderer half evaporates; #2148 shrinks to ShowField-side plumbing. |

The existing precedent points to the first: OSPRay shipped as a separate `ShowFieldWithOspray` module rather than a swap-in backend, and #2469/#2470 are named as parallel modules, not backends. If that repeats, betting #2148 on VTK is a bet that loses.

### What VTK would give you, concretely

This is worth spelling out, because the ShowField Text tab is close to a *solved problem* in VTK — it's one of its oldest built-in features:

| ShowField Text control | VTK mechanism |
|---|---|
| Show Text | add/remove actor |
| Nodes / Edges / Faces checkboxes | `vtkLabeledDataMapper` over points; `vtkCellCenters` to get edge/face anchor points |
| Text Precision | `vtkLabeledDataMapper::SetLabelFormat("%6.3g")` — printf-style |
| Text Size | `vtkTextProperty::SetFontSize` |
| Text Coloring / color picker | `vtkTextProperty::SetColor`, or per-label via a label hierarchy |
| **Cull backfacing text** | `vtkSelectVisiblePoints` — and it does true **z-buffer occlusion**, not just a backface test, so it's strictly better than the v4 behavior |
| Text always visible | omit `vtkSelectVisiblePoints` / disable actor depth test |
| *(the 60k-label scaling problem)* | `vtkPointSetToLabelHierarchy` + `vtkLabelPlacementMapper` — priority-based placement with automatic overlap decluttering and distance LOD |
| *(the font atlas)* | `vtkFreeTypeTools` — glyph caching/atlasing, already written and battle-tested |

Plus the anchoring choice comes free as a class selection rather than a shader you maintain: `vtkBillboardTextActor3D` (world-anchored, always screen-aligned and readable), `vtkTextActor3D` (rotates with the scene, can go edge-on — closest to the v4 look described above), or `vtkFollower` + `vtkVectorText` (real 3D geometry that scales with zoom).

In other words, **every item I estimated as hard is a solved, shipped VTK class.** `vtkLabeledDataMapper` alone — "label each point with its scalar value, printf-formatted, with a text property" — is essentially this entire feature.

### Revised estimate under each path

From my breakdown above:

- **Renderer-side** — atlas, billboard shader, depth/cull handling: **~4–6 days.** This is the part VTK obviates completely.
- **ShowField-side** — iterate nodes/edges/faces, format values, honor the 7 state params, un-hide the tab, `.srn5` round-trip, regression net: **~3 days.** This is renderer-agnostic and survives either path essentially unchanged. Under VTK it's mostly building a `vtkPolyData` + `vtkStringArray` instead of filling Spire buffers.

So VTK doesn't eliminate #2148; it converts a ~1.5–2 week renderer project into a ~3 day plumbing task — *conditional on* VTK landing as a backend rather than a fork, and on it landing at all.

### What I'd suggest

This does partially revise my recommendation in the previous comment. I said "land the TextBuilder atlas rework first as shared infrastructure" — that is precisely the ~4–6 days most exposed to being thrown away if VTK lands as a backend. Flagging it rather than leaving it to be discovered mid-implementation.

That said, I don't think "wait for VTK" is right either, because:

- VTK has no build integration, no code, a twice-slipped milestone, and its two biggest risks unstarted;
- #1334 (fonts inconsistent and too small across ShowString/ShowColorMap/ViewScene) is a **live user-facing complaint today**, and the per-character draw calls hurt those modules now;
- VTK is a heavyweight dependency — adding it works directly against #1487 (build time reduction), on top of the current ~15 GB / 30–90 min first build.

Reasonable middle path:

1. **Do the ~3 days of ShowField-side work regardless.** It's renderer-agnostic, it's the part that encodes actual domain knowledge (element iteration, value formatting, the coloring modes), and it ports to VTK cheaply.
2. **Hold the ~4–6 day TextBuilder atlas/billboard rework** until fork-vs-replace is settled. If VTK is a fork, or is >1 year out, do it — #1334 alone justifies it. If VTK is a genuine backend on a real schedule, skip it.
3. Settle fork-vs-replace on #2481 explicitly. It's the highest-leverage open question here, and it silently blocks more than just this issue.

One to verify if VTK proceeds: Qt compatibility. SCIRun requires Qt 5.15 or 6.3.1+, and `QVTKOpenGLNativeWidget` Qt6 support needs a recent VTK 9.x — worth pinning an exact minimum version in #2476 rather than a bare `find_package(VTK)`.

**Comment by dcwhite** [comment]
## Correction: the VTK renderer is well underway — and that *settles* the question above

My previous comment said the VTK issues were "analysis-only, no code" and that VTK "is not a dependency at all." **Both are wrong.** I checked only `master`; the work is on the **`refactor_add_vtk`** branch, where @basisunus has **51 commits from 2026-05-12 through 2026-07-28** — about eleven weeks, active as of three days ago.

Usefully, this doesn't just correct the status — it **resolves the fork-vs-replace crux I flagged as the open question**, and it resolves it in the direction that means **#2148 is not moot.**

### What's actually on the branch

**VTK is a real Superbuild external.** `Superbuild/VtkExternal.cmake` pulls `v9.6.1` from `gitlab.kitware.com/vtk/vtk.git`, wired through `Superbuild.cmake:114,261,292,330`. So #2476 is essentially done. (My Qt-compatibility caveat is moot — 9.6.1 is well past the Qt6 requirement.)

**New modules, dialogs and datatypes**, ~4,200 lines across 45 files:

| Area | Files |
|---|---|
| Renderer | `Interface/Modules/Render/Vtk/` — `VtkRenderer`, `VtkQWidget`, `VtkCameraController` |
| Modules | `Modules/Render/ViewSceneVtk.{h,cc}`, `Modules/Visualization/ShowFieldVtk.{h,cc}` |
| Dialogs | `ViewSceneVtkDialog`, `ShowFieldVtkDialog`, plus both `.ui` files |
| Factory | `Config/ViewSceneVtk.module`, `Config/ShowFieldVtk.module` |
| Data | `Core/Datatypes/VtkGeometry.{h,cc}`, `Core/Algorithms/Visualization/VtkDataAlgorithm.{h,cc}` |

The GL-context integration flagged as the top risk in #2469 has clearly been the main battle — a long run of commits on VTK/Qt context and linking, plus new `ViewQuad.vs/.fs` shaders and `es-render/TextureMan` changes, i.e. the VTK output is composited through a textured quad.

`VtkGeometryObject : public Datatype` wraps a `vtkSmartPointer<vtkDataObject>` with its own `Material`, `TransferFunc`, `BBox` and `GeometryType` — a **parallel** datatype alongside `GeometryObject`, not a conversion layer over it. There's a `CompositeVtkGeometryObject` for grouping. The most recent commit is *"added code for all geometry type conversions."*

### This settles fork-vs-replace: it's a fork

Three independent signals, all pointing the same way:

1. **`ShowFieldVtk` and `ViewSceneVtk` are separate modules** with their own `.module` configs, dialogs and `.ui` files — exactly the `ShowFieldWithOspray` precedent.
2. **`VtkRenderer` does not implement `RendererInterface`.** It's a bare `class SCISHARE VtkRenderer`, not a backend behind the abstraction. #2481's abstraction is not what this branch is built on.
3. **The branch is 4,234 insertions against 5 deletions.** Almost purely additive; the native renderer is untouched apart from CMake plumbing and a `GeometryType` enum addition.

Per the decision table in my previous comment, that lands squarely on the first row: **#2148 is not moot.** Native `ShowField` and `ViewScene` never touch VTK, so if you want text labels on the default renderer, the `TextBuilder` atlas + billboard work is still required in full. Nothing on this branch reduces it.

### And there is no text on the branch

Zero hits across it for `vtkTextActor`, `vtkTextActor3D`, `vtkBillboardTextActor3D`, `vtkLabeledDataMapper`, `vtkTextProperty`, `TextBuilder`, or `ShowText`. Text isn't part of the design yet on either side.

So the practical consequence of the fork architecture is that **text labels will need to be implemented twice** — once in `TextBuilder` for native ShowField (#2148, the ~1.5–2 weeks estimated above), and once in VTK for `ShowFieldVtk` (#2470). The VTK class inventory in my previous comment isn't wasted, it's just **filed against the wrong issue**: `vtkLabeledDataMapper` + `vtkSelectVisiblePoints` + `vtkPointSetToLabelHierarchy` are what makes labels nearly free for **#2470**, not for this one. Might be worth copying that table over there.

That duplication is a genuine cost of the fork approach and probably deserves an explicit decision rather than arriving by default — it's the same tradeoff `ShowFieldWithOspray` already made.

### Revised recommendation

This reverses the "hold" advice in my previous comment. I suggested deferring the ~4–6 day `TextBuilder` atlas/billboard rework until fork-vs-replace was settled. It is now settled, as fork — so **that work is not at risk of being superseded, and it can proceed.** The original plan from my first comment stands unchanged:

1. `TextBuilder` font atlas + batched VBO, then the billboard shader — still shared infrastructure, still also resolves #1334, still speeds up ShowString/ShowColorMap.
2. ShowField-side generation + un-hide the tab.

Two things worth raising with @basisunus while the branch is live, both cheap now and awkward later:

- **`WITH_VTK` currently defaults to `ON`** (`Superbuild.cmake:114`), which makes VTK 9.6.1 a mandatory build for everyone. Presumably a development convenience, but it should probably flip to `OFF` before merge — it's a heavyweight addition to an already ~15 GB / 30–90 min first build, and cuts directly against #1487.
- **Text/labels aren't in the `VtkGeometryObject` design yet.** If label support is wanted in `ShowFieldVtk`, it's much cheaper to leave room for it in that datatype now than to retrofit it after the geometry conversions settle.

**Comment by dcwhite** [comment]
### Follow-ups filed

Split the VTK-side observations from the review above into their own issues, all assigned to @basisunus:

- **#2606** — `WITH_VTK` defaults to `ON` on `refactor_add_vtk`, making VTK 9.6.1 mandatory for every build. Should flip to `OFF` before merge (cf. `WITH_OSPRAY`), or be an explicit documented decision. *(Build / milestone: VTK renderer)*
- **#2607** — The VTK renderer has no text rendering at all, while the native renderer draws text in three places (`ShowString`, `ShowColorMap` ticks, ViewScene axis labels). That's a parity blocker for milestone #101, "transition the main renderer to VTK." *(Graphics / milestone: Transition Main renderer to VTK)*
- **#2608** — `ShowFieldVtk` field data labels: the VTK counterpart to this issue, where `vtkLabeledDataMapper` + `vtkSelectVisiblePoints` + `vtkLabelPlacementMapper` do most of the work. *(Graphics / milestone: VTK renderer)*

**This issue keeps its full original scope** — the ~1.5–2 week `TextBuilder` atlas + billboard estimate stands. Since the VTK work is a fork rather than a backend, native `ShowField` gets no benefit from it, and #2608 does not substitute for the work here.

One caveat worth tracking: milestone #101 ("transition the main renderer to VTK instead of SR", due 2026-10-31) means the native renderer is intended to be replaced eventually, which does give the `TextBuilder` rework a shelf life. But that milestone is currently empty, is gated on a user study and testing that haven't started, and #99 (VTK renderer testing) isn't due until 2027-01-31 — *after* #101's date, so the sequencing there looks stale. Native remains the default renderer for the foreseeable future, and #1334 is a live complaint today.
